searx/engines/tubearchivist.py

```diff
@@ -0,0 +1,173 @@
+from searx.enginelib.traits import EngineTraits
+
+traits: EngineTraits
+
+about = {
+    # pylint: disable=line-too-long
+    "website": 'https://www.tubearchivist.com',
+    "official_api_documentation": 'https://github.com/tubearchivist/tubearchivist',
+    "use_official_api": True,
+    "require_api_key": False,
+    "results": 'JSON',
+}
+
+# engine dependent config
+categories = ["videos"]
+paging = True
+
+
+base_url = ""
+"""Base URL of the Tube Archivist instance.  Fill this in with your own tubearchivist url
+"""
+
+
+token = ""
+"""The API key to use for authentication.
+"""
+
+link_to_mp4 = False
+"""Optional, if true SearXNG will link directly to the mp4 of the video to play in the browser.
+The default behavior is to link into TubeArchivist's interface directly.
+"""
+
+
+def request(query, params):
+    """Assemble request for the Tubearchivist API"""
+
+    if not query:
+        return False
+
+    params['url'] = (
+        base_url.rstrip("/")
+        + "/api/search?"
+        + urlencode(
+            {
+                'query': query,
+            }
+        )
+    )
+    params['headers']['Authorization'] = f'Token {token}'
+
+    return params
+
+
+def response(resp):
+    return video_response(resp)
+
+
+def video_response(resp):
+    """Parse video response from Tubearchivist instances."""
+    results = []
+
+    json_data = resp.json()
+
+    if 'results' not in json_data:
+        return []
+
+    for channel_result in json_data['results']['channel_results']:
+        metadata = []
+        channel_url = absolute_url(f'/channel/{channel_result["channel_id"]}')
+        results.append({
+            'url': channel_url,
+            'title': channel_result['channel_name'],
+            'content': html_to_text(channel_result['channel_description']),
+            'author': channel_result['channel_name'],
+            'views': humanize_number(channel_result['channel_subs']),
+            'thumbnail': f'{absolute_url(channel_result["channel_thumb_url"])}?auth={token}',
+            'metadata': ' | '.join(metadata)
+        })
+
+
+    for video_result in json_data['results']['video_results']:
+        metadata = [
+            x
+            for x in [
+                video_result['channel']['channel_name'],
+                ' | '.join(video_result['tags']),
+            ]
+            if x
+        ]
+        if link_to_mp4:
+            url = f'{base_url.rstrip("/")}{video_result["media_url"]}'
+        else:
+            url = f'{base_url.rstrip("/")}/?videoId={video_result["youtube_id"]}'
+        results.append({
+            'url': url,
+            'title': video_result['title'],
+            'content': html_to_text(video_result['description']),
+            'author': video_result['channel']['channel_name'],
+            'length': video_result['player']['duration_str'],
+            'views': humanize_number(video_result['stats']['view_count']),
+            'template': 'videos.html',
```

Comment:
nit: please move the template definition to the beginning of the dict, so that it's visible at first glance (for consistency with all other engine implementations)

Reply:
done